lib/rexml/parsers/baseparser.rb

```diff
@@ -766,6 +771,19 @@ def process_instruction
         [:processing_instruction, name, content]
       end
 
+      if StringScanner::Version < "3.1.1"
+        def get_quote
+          @source.match(/(['"])/, true)&.[](1)
+        end
+      else
+        def get_quote
+          if quote = Private::QUOTE[@source.peek_byte]
+            @source.scan_byte
+          end
+          quote
```

Comment:
`case` may be faster than this because it creates `Hash` internally:

```suggestion
          case @source.peek_byte
          when 34 # '"'.ord
            @source.scan_byte
            '"'
          when 39 # "'".ord
            @source.scan_byte
            "'"
          else
            nil
          end
```

Reply:
* benchmark/peek_byte.yaml 
```
prelude: |
  require "strscan"
  str = "'"
  scanner = StringScanner.new(str)
  QUOTE_HASH = {'"'.ord => '"', "'".ord => "'"}
  QUOTE_ARRAY = [].tap { |x|
    x['"'.ord] = '"'
    x['"'.ord] = '"'
  }

benchmark:
  array: |
    if quote = QUOTE_ARRAY[scanner.peek_byte]
      scanner.scan_byte
    end
    quote
  hash: |
    if quote = QUOTE_HASH[scanner.peek_byte]
      scanner.scan_byte
    end
    quote
  case_single: |
    case scanner.peek_byte
    when 39 # "'".ord
      scanner.scan_byte
      "'"
    when 34 # '"'.ord
      scanner.scan_byte
      '"'
    else
      nil
    end
  case_double: |
    case scanner.peek_byte
    when 34 # '"'.ord
      scanner.scan_byte
      '"'
    when 39 # "'".ord
      scanner.scan_byte
      "'"
    else
      nil
    end
```

```
$ benchmark-driver benchmark/peek_byte.yaml 
Warming up --------------------------------------
               array    24.279M i/s -     24.496M times in 1.008947s (41.19ns/i)
                hash    22.172M i/s -     22.851M times in 1.030625s (45.10ns/i)
         case_single    25.412M i/s -     26.142M times in 1.028710s (39.35ns/i)
         case_double    25.464M i/s -     25.972M times in 1.019976s (39.27ns/i)
Calculating -------------------------------------
               array    31.871M i/s -     72.838M times in 2.285418s (31.38ns/i)
                hash    27.913M i/s -     66.515M times in 2.382942s (35.83ns/i)
         case_single    33.375M i/s -     76.236M times in 2.284205s (29.96ns/i)
         case_double    33.425M i/s -     76.391M times in 2.285431s (29.92ns/i)

Comparison:
         case_double:  33425042.8 i/s 
         case_single:  33375417.3 i/s - 1.00x  slower
               array:  31870568.1 i/s - 1.05x  slower
                hash:  27912820.8 i/s - 1.20x  slower
```

Since `case` is the fastest, we will use `case`.
Thanks!